contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+
+## Desired Approach
+
+Any valid method to update the pod spec should be applicable for vertical scaling, e.g., using kubectl commands set, patch, apply, edit.
+Logs associated with the pod will capture the failure/success of the resize command.
+The controller will continue to attempt the update to the spec while there is a difference between the current size and size in updated spec.
+If an update is partially successful, user can know this from the logs and attempt to rectify the situation by submitting new updates (that can restore original size(s) or go for a size feasible on all the nodes). 
+
+The policy for vertical scaling can itself be mutable.
+It is desirable to submit a change in policy, confirm its acceptance (by re-reading the changed pod spec) and then submit a change to a resource size impacted by that policy. 
+
+# Design and Implementation
+
+This section highlights some of implementation details by describing API changes and the workflow of vertical scaling on a pod.
+
+## Changes on API and key components
+
+This section describes briefly API changes for pod-level vertical scaling and the related changes on the 'API server', the 'Scheduler', and the 'Kubelet'.
+
+* **ResizeRequest**
+
+A new data structure, ResizeRequest, is added to v1.PodSpec:
+
+```go
+const (
+    ResizeRequested ResizeStatus = "Requested"
+    ResizeAccepted ResizeStatus = “Accepted”
+    ResizeRejected ResizeStatus = “Rejected”
+    ResizeNone      ResizeStatus = "None"
+)
+
+type ResizeRequest struct {
+    RequestStatus ResizeStatus
+    NewResources  []ResourceRequirements // indexed by containers’ index
+}
+
+Type PodSpec {
+    …
+    ResizeRequest ResizeRequest
+    ...
+}
+```
+
+ResizeRequest has two variables, RequestStatus and NewResources.
+RequestStatus represents the status of a resource resizing request.
+ResizeRequested indicates resource resizing for a pod is requested.
+ResizedAccepted and ResizedRejected means that the requested resource resizing is accepted and rejected, respectively, by the Scheduler.
+The NewResources is an array indexed by a container’s index and its each entry holds new resource requirements of a container that needs to resize.
+
+Given a new PodSpec with new resource requirements from a client, first the 'API server' validates it.
+If it is valid, the 'API server' sets the RequestStatus to ResizedRequested and copies the new resource requirements of each container into the NewResources.
+Also, the 'API server' restores the resource requirements of each container of the PodSpec to the original and writes the revised PodSpec to ETCD to communicate with the Scheduler.
+This is because at this moment the PodSpec on ETCD shouldn’t be updated with new resource requirements.
+
+For a pod with ResizeRequested, the 'Scheduler' checks if the node on which the pod currently runs has enough resources to resize the pod.
```

Comment:
This is slightly more complex. With the addition of priorities, a resource request change for a high priority pod should preempt low priority pods if needed.

Reply:
See https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/ for more details.

Reply:
Thanks for pointing it. Currently, the pod-priority isn't covered in this proposal as it was not available when we started this work, but, we're planing to support pod-priority. Basically we think that in vertical scaling, the scheduler also should follow the current policy as you mentioned. Now, I'm looking at source codes to understand how the scheduler works with pod-priority, and once we figure out how to support pod-priority in vertical scaling, we'll update the proposal accordingly.